processor/tasks/processGerbers.js

```diff
@@ -0,0 +1,248 @@
+const fs = require('fs')
+const globule = require('globule')
+const path = require('path')
+const cp = require('child_process')
+const Jszip = require('jszip')
+const util = require('util')
+
+const { exists } = require('../utils')
+const exec = util.promisify(cp.exec)
+const writeFile = util.promisify(fs.writeFile)
+const readFile = util.promisify(fs.readFile)
+
+const gerberFiles = require('./gerber_files')
+const boardBuilder = require('./board_builder')
+
+const topSvgPath = 'images/top.svg'
+const bottomSvgPath = 'images/bottom.svg'
+const zipInfoPath = 'zip-info.json'
+const topPngPath = 'images/top.png'
+const topLargePngPath = 'images/top-large.png'
+const topMetaPngPath = 'images/top-meta.png'
+const topWithBgndPath = 'images/top-with-background.png'
+
+async function processGerbers(
+  eventEmitter,
+  inputDir,
+  kitspaceYaml,
+  outputDir,
+  hash,
+  name,
+) {
+  const zipPath = name.split('/')[1] + '-' + hash.slice(0, 7) + '-gerbers.zip'
+
+  eventEmitter.emit('in_progress', topSvgPath)
+  eventEmitter.emit('in_progress', bottomSvgPath)
+  eventEmitter.emit('in_progress', zipPath)
+  eventEmitter.emit('in_progress', zipInfoPath)
+  eventEmitter.emit('in_progress', topPngPath)
+  eventEmitter.emit('in_progress', topLargePngPath)
+  eventEmitter.emit('in_progress', topMetaPngPath)
+  eventEmitter.emit('in_progress', topWithBgndPath)
+
+  const failAll = e => {
+    eventEmitter.emit('failed', topSvgPath, e)
+    eventEmitter.emit('failed', bottomSvgPath, e)
+    eventEmitter.emit('failed', zipPath, e)
+    eventEmitter.emit('failed', zipInfoPath, e)
+    eventEmitter.emit('failed', topPngPath, e)
+    eventEmitter.emit('failed', topLargePngPath, e)
+    eventEmitter.emit('failed', topMetaPngPath, e)
+    eventEmitter.emit('failed', topWithBgndPath, e)
+  }
+
+  try {
+    const gerberDir = kitspaceYaml.gerbers || ''
+    const color = kitspaceYaml.color || 'green'
+
+    await exec('mkdir -p ' + path.join(outputDir, 'images'))
+
+    if (
+      !(await exists(path.join(outputDir, zipPath))) ||
+      !(await exists(path.join(outputDir, bottomSvgPath))) ||
+      !(await exists(path.join(outputDir, zipInfoPath))) ||
+      !(await exists(path.join(outputDir, topSvgPath))) ||
+      !(await exists(path.join(outputDir, topPngPath))) ||
+      !(await exists(path.join(outputDir, topLargePngPath))) ||
+      !(await exists(path.join(outputDir, topMetaPngPath))) ||
+      !(await exists(path.join(outputDir, topWithBgndPath)))
+    ) {
```

Comment:
Can be improved improved for sure, though I am not sure how yet. Accessing these paths separately does give us a bit of control that I might like to keep. Not sure how to clean it up while keeping it. Using DeMorgan's doesn't make it more readable in my view, neither would a seperate function (`pathsExist(outputDir, [zipPath, bottomSvgPath, zipInfoPath ...`)

Reply:
> neither would a separate function (`pathsExist(outputDir, [zipPath, bottomSvgPath, zipInfoPath ...`)

I meant a function named after the of this condition, e.g., `if (!allAssetsPathsExist())` just like this and  `allAssetsPathsExist` is responsible for accessing the files.


> using DeMorgan's doesn't make it more readable in my view

It can be read: all the conditions inside the parentheses must be `true` to make the whole expression `true`. With the negation, which is the case of this snippet, if any of the conditions inside the parentheses is false the condition will be true.

Reply:
Happy? ;)

https://github.com/kitspace/kitspace-v2/blob/7d4627211163db4ea83071c85d6aeff2479618e3/processor/tasks/processGerbers/index.js#L49-L54